src/rrdp-iostat/rrdp_iostat.ml

```diff
@@ -646,12 +667,12 @@ module Iostats_value = struct
         avgqu_sz = avgqu_sz /. 5.;
       }
 
-  let sumup (values : t list) : t =
-    List.fold_left (fun acc v ->
-        {
-          latency = acc.latency +. v.latency;
-          avgqu_sz = acc.avgqu_sz +. v.avgqu_sz;
-        }) empty values
+  let accumulate (values : t list) : t =
+    let max acc v = { latency = Float.max acc.latency v.latency
+                    ; avgqu_sz = Float.max acc.avgqu_sz v.avgqu_sz
```

Comment:
Latency is correct here, but I'm not sure what the best aggregation function is for queue size. If each of the N VMs has a queue size of 1, then in total you do have N active requests, so perhaps sum is right for queue size
However those requests could have all come in at different times, so equally a queue size of 1 globally may be accurate, or any number between 1 and N: we only measure queue size per VM, I don't think we can use that to calculate queue size per device, we simply don't have enough information about the timings to determine that.

Can't we get this information from iostats instead?

Reply:
this code is in the `Iostat_value` module, so `avgque_sz` is already obtained from `iostats -x`

to me it makes sense if we take the max, since the user is probably more interested in the worst average queue size

Reply:
@edwintorok are you happy with this?

Reply:
I don't think we have enough information to calculate the average queue size, perhaps we should just report it as NaN instead?

Consider this:
VM1: IO, IDLE, IO, IDLE, IO, IDLE.  queue size = 1
VM2: IDLE, IO, IDLE, IO, IDLE, IO.  queue size = 1
Total  queue size: 1. With + you used to get 2, incorrect. With max you get 1: correct.

Now consider this:
VM1: IO, IO(2), IO(3), IDLE, IO, IO(2), IO(3), IDLE.  queue size = 3
VM2: IDLE, IO, IDLE, IO, IDLE, IO, IDLE  queue size = 1
Total  queue size: 3. With + you used to get 4, incorrect. With max you get 3: correct.

A slight variation:
VM1: IO, IO(2), IO(3), IDLE, IO, IO(2), IO(3), IDLE.  queue size = 3
VM2: IO, IDLE, IO, IDLE, IO, IDLE, IO  queue size = 1
Total  queue size: 4 (you got IO(3) overlap with IO). With + you used to get 4, correct. With max you get 3: incorrect.

Reply:
The only way to know true total queue size is to track that at the storage block device level and ask it for the total, we can't infer it from the individual averages, because we don't know how these averages have overlapped.
That is why I was asking whether we can use iostat to query the total queue size on the block device, instead of on the individual VM disks (although in general I think that is not possible, because the device may be on NFS, etc. which has no block devices where we could total these things up).

Reply:
i would be in favour of not displaying the metric

we could make this slightly better by returning the queue size when there is only 1 VM, and NaN when there is more than one

Reply:
i think we might be overthinking this a little. if i understand correctly:
 -  `iostat -x` reports based on devices in `/dev`, in particular it gives us `avgqu_sz` per SR
- in the rrdd_iostat plugin we accumulate `avgqu_sz` from `iostat -x` over the VDIs, for each SR, to create an `avgqu_sz` *datasource* for each SR

but since `iostat -x` gives output per SR, not per VDI, the `avgqu_sz` for every VDI on a single SR should be identical - i.e. `sum` is wrong, `max` is correct, but we don't actually need to accumulate at all - all we need is information about one VDI

(for future me: this helped me understand what iostat is doing a bit better https://www.xaprb.com/blog/2010/01/09/how-linux-iostat-computes-its-results/)

Reply:
an NFS SR won't have a device in /dev, you would only have the individual devices. Lets merge the PR for now and think about what to do about the invalid value later.

Reply:
yes you're right, i must have been mistaken when i was looking at it earlier this week - i tried this and i did see an increase in the number of devices

i created a new PR for fixing this: https://github.com/xenserver/rrdd-plugins/pull/62, so we can move the discussion there